src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -80,11 +81,37 @@ public CursorsService(final ZooKeeperHolder zkHolder,
 
         validateStreamId(cursors, streamId, subscriptionId);
 
+        LOG.debug("[COMMIT_CURSORS] stream IDs validation finished");
+
+        final Iterator<EventTypePartition> distinctPartitions = cursors.stream()
+                .map(c -> new EventTypePartition(c.getEventType(), c.getPartition()))
+                .distinct()
+                .iterator();
+
+        final HashMap<EventTypePartition, Iterator<Boolean>> partitionCommits = new HashMap<>();
+        while (distinctPartitions.hasNext()) {
+
+            final EventTypePartition etPartition = distinctPartitions.next();
+            final List<SubscriptionCursor> partitionCursors = cursors.stream()
+                    .filter(etPartition::ownsCursor)
+                    .collect(Collectors.toList());
+
+            final Iterator<Boolean> commitResultIterator = processPartitionCursors(subscriptionId, partitionCursors,
```

Comment:
I need the iterator here because I need to fill the result list in the same order as user provided cursors. For that I need to iterate through cursors and and for each cursor get a result from appropriate list depending on EventTypePartition. Using iterators was the easiest way to implement that because iterators keep iteration pointer.